### PR TITLE
Flutter run restart

### PR DESCRIPTION
**Pull request by devoncarew**
- add a `restartApp()` method to Device
- have the adb implementation of it use the logic from `flutter refresh`
- use the new `Flutter.FrameworkInitialization` event to measure our startup time

The edit-refresh time should drop from ~7s to ~2s.

### CODE REVIEW

**Comment by devoncarew** [comment]
@jason-simmons @yjbanov for review please :)

**Comment by jason-simmons** [comment]
lgtm

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 1]
```diff
@@ -355,6 +358,39 @@ class AndroidDevice extends Device {
   }
 
   @override
+  Future<bool> restartApp(
+    ApplicationPackage package,
+    LaunchResult result, {
+    String mainPath,
+    Observatory observatory
+  }) async {
+    Directory tempDir = await Directory.systemTemp.createTemp('flutter_tools');
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/android/android_device.dart]: Shouldn't all build artifacts go in `build` for easier debugging?

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 2]
```diff
@@ -433,18 +442,31 @@ class _RunAndStayResident {
...
-      if (observatory != null && !observatory.isClosed && _isolateId != null) {
-        observatory.flutterExit(_isolateId);
-
-        // WebSockets do not have a flush() method.
-        await new Future<Null>.delayed(new Duration(milliseconds: 100));
+      try {
+        if (observatory != null && !observatory.isClosed) {
+          if (observatory.isolates.isNotEmpty) {
+            observatory.flutterExit(observatory.firstIsolateId);
+            // The Dart WebSockets API does not have a flush() method.
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/commands/run.dart]: Is there a bug we can track? Maybe provide a link to the relevant SDK github issue?

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 3]
```diff
@@ -556,6 +557,21 @@ class IOSSimulator extends Device {
   }
 
   @override
+  Future<bool> restartApp(
+    ApplicationPackage package,
+    LaunchResult result, {
+    String mainPath,
+    Observatory observatory
+  }) async {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: This doesn't need to be `async` as it doesn't `await` on anything.

Comment (yjbanov) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: Alternatively, you would rewrite it using `await` and `try`/`catch`.

File: packages/flutter_tools/lib/src/observatory.dart [hunk 4]
```diff
@@ -54,16 +74,30 @@ class Observatory {
     _getEventController(data['streamId']).add(event);
   }
 
+  Future<IsolateRef> waitHasIsolate() {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/observatory.dart]: The name does not imply that it returns the first isolate when it becomes available. How about this:

```dart
Future<IsolateRef> get firstIsolate {
  ..
}
```

File: packages/flutter_tools/lib/src/observatory.dart [hunk 5]
```diff
@@ -54,16 +74,30 @@ class Observatory {
...
+    if (isolates.isNotEmpty) {
+      return new Future<IsolateRef>.value(isolates.first);
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/observatory.dart]: Using `async`/`await` you could avoid futures and `.then`.

File: packages/flutter_tools/lib/src/observatory.dart [hunk 6]
```diff
@@ -113,18 +160,35 @@ class Response {
...
-  Event(this.event);
+class Event extends Response {
+  Event(Map<String, dynamic> response) : super(response);
 
-  final Map<String, dynamic> event;
+  String get kind => response['kind'];
+  IsolateRef get isolate => new IsolateRef.from(response['isolate']);
 
-  String get kind => event['kind'];
+  /// Only valid for [kind] == `Extension`.
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/observatory.dart]: Does `///` in inline comments do anything useful?

Comment (yjbanov) [inline packages/flutter_tools/lib/src/observatory.dart]: nvm, I misread the code.

**Comment by yjbanov** [comment]
lgtm

**Comment by yjbanov** [comment]
We need a good testing strategy for `flutter_tools`.

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_device.dart]: This is the pattern that the current `flutter refresh` command uses. We may want to put more things into `build/`, and use a temp dir less often. I'd still want most of the build artifacts in `build/` to be semantic wrt to the user - something that they would understand and identify as useful to their project.

File: packages/flutter_tools/lib/src/commands/run.dart (hunk 2 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/run.dart]: filed https://github.com/dart-lang/sdk/issues/26547

File: packages/flutter_tools/lib/src/ios/simulators.dart (hunk 3 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: removed async

File: packages/flutter_tools/lib/src/observatory.dart (hunk 4 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/observatory.dart]: `get waitFirstIsolate`?

File: packages/flutter_tools/lib/src/observatory.dart (hunk 5 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/observatory.dart]: updated

File: packages/flutter_tools/lib/src/observatory.dart (hunk 4 above)
Comment (yjbanov) [inline packages/flutter_tools/lib/src/observatory.dart]: I think `Future` implies the waiting, but sure.

**Pull request merged**